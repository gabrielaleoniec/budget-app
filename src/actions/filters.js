export const setTextFilter = (text = '') => ({
    type: "SET_TEXT_FILTER",
    text
});

export const sortByDate = (sortByAsc = false) => ({
    type: "SORT_BY",
    sortBy: 'createdAt', 
    sortByAsc
});

export const sortByAmount = (sortByAsc = false) => ({
    type: "SORT_BY",
    sortBy: 'amount',
    sortByAsc
});

export const setStartDate = (startDateString = '') => {
    const startDate = new Date(startDateString).valueOf();
    return {
        type: "SET_START_DATE",
        startDate: isNaN(startDate) ? undefined : startDate
    }
}

export const setEndDate = (endDateString = '') => {
    const endDate = new Date(endDateString);
    const dateAsMiliseconds = Date.parse(endDate);

    return {
        type: "SET_END_DATE",
        endDate: isNaN(dateAsMiliseconds) ? undefined : dateAsMiliseconds
    }
}