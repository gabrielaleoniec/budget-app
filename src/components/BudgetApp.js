import React from 'react';

const BudgetApp = () => {
    return <div>Inside the compppponent</div>
}

export default BudgetApp;