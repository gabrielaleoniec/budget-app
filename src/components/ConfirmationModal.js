import React from 'react';

const ConfirmationModal = (props) => {
    const className = props.showConfirmationBox ? "confirmation-box--show" : '';
    return <div className="box-layout box-layout--cover">
        <div className={`confirmation-box ${className}`}>
            <p>Should the expense be removed?</p>
            <div className="confirmation-box__buttons">
                <button className="button button--weak" onClick={props.hideConfirmationModal}>Cancel</button>
                <button className="button button--remove" onClick={props.removeExpense}>Remove</button>
            </div>
        </div>
    </div>
}

export default ConfirmationModal;
