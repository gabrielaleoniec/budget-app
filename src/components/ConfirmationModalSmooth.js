import React from 'react';

const ConfirmationModalSmooth = () => {
    return () => {
        
    }
}