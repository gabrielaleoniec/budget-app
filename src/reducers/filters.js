import moment from 'moment';

const defaultStateFilterReducer = {
    text: '',
    sortBy: 'createdAt',
    sortByAsc: false,
    startDate: moment().startOf('month').valueOf(),
    endDate: moment().endOf('month').valueOf()
}

const filterReducer = (state = defaultStateFilterReducer, action) => {
    switch (action.type) {
        case 'SET_TEXT_FILTER':
            return {
                ...state,
                text: action.text
            }
        case 'SORT_BY':
            console.log('reducer', action);
            return {
                ...state,
                sortBy: action.sortBy,
                sortByAsc: action.sortByAsc
            }
        case 'SET_START_DATE':
            return {
                ...state,
                startDate: action.startDate
            }
        case 'SET_END_DATE':
            return {
                ...state,
                endDate: action.endDate
            }
        default:
            return state;
    }
}

export default filterReducer;