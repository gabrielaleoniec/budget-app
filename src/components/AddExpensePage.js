import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

const AddExpensePage = (props) => (
    <div>
        <h2>Add expense</h2>
        <ExpenseForm onSubmit={(expense) => {
            props.dispatch(addExpense(expense));
            props.history.push('/');
        }}/>
    </div>
);

const mapStateToProps = (state) => ({
    expenses: state.expenses
})

export default connect(mapStateToProps)(AddExpensePage);