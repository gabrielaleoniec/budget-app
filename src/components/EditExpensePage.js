import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';

import ExpenseForm from './ExpenseForm';
import Confirmation from './ConfirmationModal';
import { startUpdateExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            showConfirmation: false,
            showConfirmationBox: false,
            time: 500
        }
    }
    onSubmit = (expense) => {
        const createdAt = typeof expense.createdAt === 'string' ? moment(expense.createdAt).valueOf() : expense.createdAt;

        this.props.startUpdateExpense(this.props.expense.id, { ...expense, createdAt });
        this.props.history.push('/');
    }

    onRemove = () => {
        this.props.startRemoveExpense({ id: this.props.expense.id });
        this.props.history.push('/');
    }

    showConfirmationModal = () => {
        this.setState({
            showConfirmation: true
        })

        this.openTimeout = setTimeout(() => {
            this.setState({
                showConfirmationBox: true
            })
        }, this.state.time);
    }

    hideConfirmationModal = () => {
        this.setState({
            showConfirmationBox: false
        })

        this.openTimeout = setTimeout(() => {
            this.setState({
                showConfirmation : false
            })
        }, this.state.time);
    }

    addClassname = () => {

    }

    render() {
        if (!this.props.expense || !this.props.expense.id) {
            return <p>Expense for the given id does not exist</p>
        }

        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h2 className="page-header__title">Edit expense</h2>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm onSubmit={this.onSubmit} expense={this.props.expense} />
                    <button onClick={this.showConfirmationModal} className="button button--remove">Remove expense</button>
                    {(this.state.showConfirmation || this.state.showConfirmationBox) && <Confirmation
                        removeExpense={this.onRemove}
                        hideConfirmationModal={this.hideConfirmationModal}
                        time={this.state.time}
                        showConfirmationBox={this.state.showConfirmationBox}
                    />}
                </div>
            </div>
        )
    }
}
const mapDispatchToProps = (dispatch, props) => ({
    startUpdateExpense: (id, expense) => dispatch(startUpdateExpense(id, expense)),
    startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
})

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);