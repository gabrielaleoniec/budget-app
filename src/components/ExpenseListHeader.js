import React from 'react';
import { connect } from 'react-redux';
import { sortByDate, sortByAmount } from '../actions/filters';

const onSortByDate = (sortByDate, asc) => {
    sortByDate(!asc)
}

export const ExpenseListHeader = ({ filters, sortByDate, sortByAmount }) => {
    return <div className="list-header">
        <div className="show-for-mobile">Expenses</div>
        <div className="show-for-desktop list-header__title">Expense</div>
        <div className="show-for-desktop list-header__date">Date
            <button
                className={`button button--sort ${filters.sortBy == 'createdAt' && (filters.sortByAsc ? 'button--top' : 'button--down')}`}
                onClick={() => {
                    sortByDate(!filters.sortByAsc);
                }}>
                <span className="sr-only">Sort by date</span>
            </button>
        </div>
        <div className="show-for-desktop list-header__amount">Amount
            <button
                className={`button button--sort ${filters.sortBy == 'amount' && (filters.sortByAsc ? 'button--top' : 'button--down')}`}
                onClick={() => {
                    sortByAmount(!filters.sortByAsc);
                }}>
                <span className="sr-only">Sort by amount</span>
            </button>
        </div>
    </div>
}

const mapStateToProps = (state) => ({
    filters: state.filters
})

const mapDispatchToProps = (dispatch) => ({
    sortByDate: (asc) => dispatch(sortByDate(asc)),
    sortByAmount: (asc) => dispatch(sortByAmount(asc))
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListHeader);