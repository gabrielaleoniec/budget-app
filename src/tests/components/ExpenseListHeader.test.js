import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseListHeader } from '../../components/ExpenseListHeader';
import { filters, altFilters } from '../fixtures/filters';

let sortByDate, sortByAmount, wrapper;
beforeEach(() => {
    sortByDate = jest.fn();
    sortByAmount = jest.fn();
    wrapper = shallow(
        <ExpenseListHeader
            filters={filters}
            sortByDate={sortByDate}
            sortByAmount={sortByAmount}
        />);
});

test('Should render ExpenseListHeader correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('Should handle sort by date', () => {
    wrapper.setProps({
        filters: altFilters
    })
    wrapper.find('button').first().simulate('click');
    expect(sortByDate).toHaveBeenCalled();
});

test('Should handle sort by amount', () => {
    wrapper.setProps({
        filters: altFilters
    })
    wrapper.find('button').last().simulate('click');
    expect(sortByAmount).toHaveBeenCalled();
});