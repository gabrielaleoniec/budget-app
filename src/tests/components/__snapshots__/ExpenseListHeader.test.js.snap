// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseListHeader correctly 1`] = `
<div
  className="list-header"
>
  <div
    className="show-for-mobile"
  >
    Expenses
  </div>
  <div
    className="show-for-desktop list-header__title"
  >
    Expense
  </div>
  <div
    className="show-for-desktop list-header__date"
  >
    Date
    <button
      className="button button--sort button--down"
      onClick={[Function]}
    >
      <span
        className="sr-only"
      >
        Sort by date
      </span>
    </button>
  </div>
  <div
    className="show-for-desktop list-header__amount"
  >
    Amount
    <button
      className="button button--sort false"
      onClick={[Function]}
    >
      <span
        className="sr-only"
      >
        Sort by amount
      </span>
    </button>
  </div>
</div>
`;
